Copyright © 2023, United States Government, as represented by the Administrator of the National Aeronautics and Space Administration. All rights reserved. 
The NASA Ames Mars Global Climate Model Patches is licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE‐2.0.
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
diff --git a/src/plevel/plev_constants.F90 b/src/plevel/plev_constants.F90
index e07ec96..ebc2493 100644
--- a/src/plevel/plev_constants.F90
+++ b/src/plevel/plev_constants.F90
@@ -25,9 +25,24 @@ module plev_constants_mod
 implicit none
 private
 
+#ifndef MARS
 real, public, parameter :: GRAV   = 9.80
 real, public, parameter :: RDGAS  = 287.04
 real, public, parameter :: RVGAS  = 461.50
 real, public, parameter :: TFREEZE  = 273.16
+#else
+
+real, public, parameter :: RADIUS = 3386.0e3
+real, public, parameter :: OMEGA  = 7.0882e-5
+real, public, parameter :: GRAV   = 3.72
+real, public, parameter :: RDGAS  = 189.02
+real, public, parameter :: CP_AIR = 735.94
+real, public, parameter :: KAPPA  = RDGAS/CP_AIR
+real, public, parameter :: RVGAS = 461.50
+real, public, parameter :: TFREEZE  = 273.16
+
+#endif
+
+
 
 end module plev_constants_mod
diff --git a/src/plevel/plevel.sh.in b/src/plevel/plevel.sh.in
index 38bd4fd..f6a3e19 100755
--- a/src/plevel/plevel.sh.in
+++ b/src/plevel/plevel.sh.in
@@ -48,6 +48,7 @@ allow_zero=.false.
 default_missval=.false.
 mask_extrap=.true.
 tlist=
+pk_name="pk"
 
 # default is ncep reanalysis levels
 plevs="100000 92500 85000 70000 60000 50000 40000 30000 25000 \
@@ -92,6 +93,8 @@ are "bk", "pk", and "ps".
                 to input file settings.
     -s 1|0      If using NetCDF4, use shuffle if 1 and don't use if 0.
                 Defaults to input file settings.
+    -k name     Name for pressure part of hybrid sigma in history file.
+                Defaults to "pk"
     -h          Display this help and exit.
     -V          Output version information and exit
 
@@ -108,7 +111,7 @@ EOF
 }
 #-----------------------------------------------------------------------
 
-while getopts 03amxi:o:d:p:t:z:s:hV arg
+while getopts 03amxi:o:d:p:t:z:s:k:hV arg
 do
     case $arg in
         0) allow_zero=.true.;;
@@ -123,6 +126,7 @@ do
         t) tlist=$OPTARG;;
         z) deflation=$OPTARG;;
         s) shuffle=$OPTARG;;
+        k) pk_name=$OPTARG;;
         h) usage; exit 0;;
         V)
             $pkg_libexec_dir/print_version $script_name
@@ -232,6 +236,10 @@ done
 
 echo " /" >> $namelist
 
+echo " &names" >> $namelist
+echo "    pk_name = $pk_name" >> $namelist
+echo " /" >> $namelist
+
 eval $plevel_interp
 if [ $? -eq 0 ]
 then
diff --git a/src/plevel/pressure_interp.F90 b/src/plevel/pressure_interp.F90
index 51c9ecf..f786ecf 100644
--- a/src/plevel/pressure_interp.F90
+++ b/src/plevel/pressure_interp.F90
@@ -102,10 +102,10 @@ contains
          k = index(i,j)
          factr(i,j) = (log_pout          -log_pfull(i,j,k)) /   &
                       (log_pfull(i,j,k-1)-log_pfull(i,j,k))
-        !factr(i,j) = min(1.,factr(i,j)) ! no extrap above top level
-        !factr(i,j) = max(0.,factr(i,j)) ! no extrap below bottom level
-         factr(i,j) = min(+1.5,factr(i,j)) ! limit extrap above top level
-         factr(i,j) = max(-0.5,factr(i,j)) ! limit extrap below bottom level
+        factr(i,j) = min(1.,factr(i,j)) ! no extrap above top level
+        factr(i,j) = max(0.,factr(i,j)) ! no extrap below bottom level
+        ! factr(i,j) = min(+1.5,factr(i,j)) ! limit extrap above top level
+        ! factr(i,j) = max(-0.5,factr(i,j)) ! limit extrap below bottom level
          up(i,j,n) = u(i,j,k) + factr(i,j) * (u(i,j,k-1) - u(i,j,k))
          vp(i,j,n) = v(i,j,k) + factr(i,j) * (v(i,j,k-1) - v(i,j,k))
       enddo
@@ -191,7 +191,10 @@ contains
 
          do j = 1, jdim
          do i = 1, idim
-             do k = 2, kbot(i,j)
+             if (log_pout(n) < log_pfull(i,j,1)) then
+                exit
+             endif
+             do k = 2,kbot(i,j)
                 if (log_pout(n) <= log_pfull(i,j,k)) then
                     Control % index (i,j,n) = k
                     Control % mask  (i,j,n) = .false.
@@ -207,10 +210,10 @@ contains
              Control % factr(i,j,n) =  &
                                (log_pout(n) - log_pfull(i,j,k)) / &
                         (log_pfull(i,j,k-1) - log_pfull(i,j,k))
-            !Control % factr(i,j,n) = min(1.,Control%factr(i,j,n)) ! no extrap above top level
-            !Control % factr(i,j,n) = max(0.,Control%factr(i,j,n)) ! no extrap below bottom level
-             Control % factr(i,j,n) = min(+1.5,Control%factr(i,j,n)) ! limit extrap above top level
-             Control % factr(i,j,n) = max(-0.5,Control%factr(i,j,n)) ! limit extrap below bottom level
+            Control % factr(i,j,n) = min(1.,Control%factr(i,j,n)) ! no extrap above top level
+            Control % factr(i,j,n) = max(0.,Control%factr(i,j,n)) ! no extrap below bottom level
+            ! Control % factr(i,j,n) = min(+1.5,Control%factr(i,j,n)) ! limit extrap above top level
+            ! Control % factr(i,j,n) = max(-0.5,Control%factr(i,j,n)) ! limit extrap below bottom level
          enddo
          enddo 
 
diff --git a/src/plevel/run_pressure_interp.F90 b/src/plevel/run_pressure_interp.F90
index 881f011..5b90706 100644
--- a/src/plevel/run_pressure_interp.F90
+++ b/src/plevel/run_pressure_interp.F90
@@ -130,9 +130,10 @@ program run_pressure_interp
       & pk_name = 'pk',&
       & res_name = 'res',&
       & hght_name = 'hght',&
-      & slp_name = 'slp'
+      & slp_name = 'slp',&
+      & tod_name = 'time_of_day_24'
 
-   character(len=NF90_MAX_NAME) :: pres_name_out = 'level'
+   character(len=NF90_MAX_NAME) :: pres_name_out = 'pstd'
    character(len=NF90_MAX_NAME), dimension(MAX_FIELD_NAMES) :: field_names
 
    type field_type
@@ -180,7 +181,9 @@ program run_pressure_interp
    logical :: allow_zero_topog = .false.
    logical :: allow_zero_sphum = .false.
 
-   integer :: time_beg=1, time_end=0, time_inc=1
+integer :: time_beg=1, time_end=0, time_inc=1
+!rjw
+real    :: missing_value_inpt = -1.e10
 
    ! user-requested deflation and shuffle settings, -1 means use input file settings
    integer :: user_deflation=-1, user_shuffle=-1
@@ -202,14 +205,16 @@ program run_pressure_interp
 
    !----- axis names -----
    character(len=NF90_MAX_NAME) :: xmass_in, ymass_in
-   character(len=NF90_MAX_NAME) :: pfull_in, time_in
+   character(len=NF90_MAX_NAME) :: pfull_in, time_in,tod_in
+
+   logical :: do_diurn = .false.
 
    character(len=NF90_MAX_NAME) :: xmass_out, ymass_out
    character(len=NF90_MAX_NAME) :: edge_name
    character(len=NF90_MAX_NAME) :: attname
    character(len=NF90_MAX_NAME) :: file_title
 
-   integer :: dimids_mass(4)
+   integer :: dimids_mass(5)
 
    integer :: ncid_in, ncid_out
    type(pres_interp_type) :: Interp(2)
@@ -242,19 +247,21 @@ program run_pressure_interp
    integer :: varid_hght_out, varid_slp_out
 
    integer :: varid_time_in,        varid_time_out
+   integer :: varid_tod_in,       varid_tod_out
    integer :: varid_time_bounds_in, varid_time_bounds_out
    integer :: varid_tavg_in(3),     varid_tavg_out(3)
 
    integer :: index_ps, index_zsurf, index_temp, index_sphum
 
    integer :: istat, ivar, itime, otime, i, n, kv, attnum, numatts, recdim,&
-      & nlon, nlat, nlev, ndim, dimids(NF90_MAX_VAR_DIMS), dimid, naxes, dimid_tbnds
+   & nlon, nlat, nlev, ndim, dimids(NF90_MAX_VAR_DIMS), dimid, naxes, dimid_tbnds, &
+   & ntod, itod, todid
    integer :: attlen
-   integer :: nlev_out, start(4), kount(4), ntimes, num_field_names
+   integer :: nlev_out, start(5), kount(5), ntimes, num_field_names
    integer :: itime_beg, itime_end, itime_inc
    integer :: isg, ieg, jsg, jeg, is, ie, js, je
    real    :: mval_hght, mval_slp
-   real(8) :: time, time_bnds(2)
+   real(8) :: time, time_bnds(2),tod
    logical :: done, do_phalf
    integer :: idim, nlen
    integer :: unit, ierr, io, in_format, cmode
@@ -263,6 +270,7 @@ program run_pressure_interp
    real, allocatable, dimension(:)     :: bk, pk
    real, allocatable, dimension(:,:)   :: res
    real, allocatable, dimension(:,:,:) :: zsurf, ps
+   real, allocatable, dimension(:,:,:) :: psdiurn
 
    real, parameter :: tlapse = 6.5e-3
    real, parameter :: tref   = 288.
@@ -286,11 +294,12 @@ program run_pressure_interp
    if (exists) then
       open (10, file='plev.input.nml')
       read (10, nml=input, end=1, err=1)
+      read (10, nml=names, end=1, err=1)
 1     close (10)
    endif
 
    do i = 1, MAX_PLEVS
-      if (pout(i) < 1.e-6) exit
+      if (pout(i) < 1.e-10) exit
       nlev_out = i
    enddo
 
@@ -370,20 +379,38 @@ program run_pressure_interp
    istat = NF90_INQUIRE_VARIABLE (ncid_in, varid_ps_in, ndims=ndim, dimids=dimids)
    if (istat /= NF90_NOERR) call error_handler (ncode=istat)
 
-   if (ndim /= 3) call error_handler ('variable '//trim(ps_name)//&
-      & ' must have three dimensions: x,y,t')
+   if (ndim /= 3) then
+      istat = NF90_INQ_DIMID (ncid_in,tod_name,todid)
+      if (istat /= NF90_NOERR) then
+         tod_name='time_of_day_12'
+         istat = NF90_INQ_DIMID (ncid_in,tod_name,todid)
+         if (istat /= NF90_NOERR) call error_handler(trim(ps_name)//' has more than three dimensions ' &
+            //' not including time of day 12 or 24', ncode=istat)
+      endif
+      do_diurn = .true.
+      print*,'tod_name is ',tod_name, 'dim ids = ',dimids(1:ndim), 'todid = ',todid
+   else
+      call error_handler ('variable '//trim(ps_name)//&
+         & ' must have three dimensions: x,y,t')
+   endif
 
    ! size and name of input horizontal mass axes
    istat = NF90_INQUIRE_DIMENSION (ncid_in, dimids(1), name=xmass_in, len=nlon)
    if (istat /= NF90_NOERR) call error_handler ('getting name and length of first mass dimension', ncode=istat)
    istat = NF90_INQUIRE_DIMENSION (ncid_in, dimids(2), name=ymass_in, len=nlat)
    if (istat /= NF90_NOERR) call error_handler ('getting name and length of second mass dimension', ncode=istat)
-   istat = NF90_INQUIRE_DIMENSION (ncid_in, dimids(3), name=time_in)
+   istat = NF90_INQUIRE_DIMENSION (ncid_in, recdim, name=time_in)
    if (istat /= NF90_NOERR) call error_handler ('getting name of last mass dimension (time)', ncode=istat)
 
+   if (do_diurn) then
+      istat = NF90_INQUIRE_DIMENSION ( ncid_in, todid, name=tod_in, len=ntod)
+      if (istat /= NF90_NOERR) call error_handler ('getting time of day mass dimension', ncode=istat)
+      print*,'ntod = ',ntod,'name tod = ',trim(tod_in),'todid = ',todid
+   endif
+
    ! set output fields using all input fields
    if (do_all_fields .or. do_all_3d_fields) then
-      call set_all_field_names ( ncid_in, field_names, num_field_names )
+      call set_all_field_names ( ncid_in, do_diurn, field_names, num_field_names )
    endif
 
    ! error: if there were no requested fields
@@ -410,6 +437,7 @@ program run_pressure_interp
    allocate (res(is:ie,js:je),&
       & zsurf(is:ie,js:je,1),&
       & ps(is:ie,js:je,1))
+   allocate (psdiurn(is:ie,js:je,1))
 
    ! read vert coord variables and set up vert coord
    istat = NF90_GET_VAR (ncid_in, varid_bk, bk)
@@ -518,7 +546,9 @@ program run_pressure_interp
    istat = NF90_DEF_VAR (ncid_out, trim(pres_name_out), NF90_REAL4, (/dimids_mass(3)/), varid_pout)
    if (istat /= NF90_NOERR) call error_handler (ncode=istat)
    ! attributes for this axis
-   istat = NF90_PUT_ATT (ncid_out, varid_pout, 'units', 'hPa')
+   !     istat = NF90_PUT_ATT (ncid_out, varid_pout, 'units', 'hPa')
+   !rjw          Output pressure grid in Pa rather than hPa
+   istat = NF90_PUT_ATT (ncid_out, varid_pout, 'units', 'Pa')
    if (istat /= NF90_NOERR) call error_handler ('putting attributes for pout dimension', ncode=istat)
    istat = NF90_PUT_ATT (ncid_out, varid_pout, 'long_name', 'pressure')
    if (istat /= NF90_NOERR) call error_handler (ncode=istat)
@@ -530,9 +560,17 @@ program run_pressure_interp
    if (istat /= NF90_NOERR) call error_handler (ncode=istat)
 
    ! copy the time axis
-   ! get variable identifiers and length
-   call copy_axis (ncid_in, time_in, ncid_out, dimids_mass(4),&
-        & varid_time_in, varid_time_out, ntimes)
+   ! get time of day information
+   if (do_diurn) then
+      call copy_axis(ncid_in, time_in, ncid_out, dimids_mass(5), &
+           & varid_time_in, varid_time_out, ntimes)
+      call copy_axis(ncid_in, tod_in, ncid_out, dimids_mass(4), &
+           & varid_tod_in, varid_tod_out, ntod)
+   else
+      ! get variable identifiers and length
+      call copy_axis (ncid_in, time_in, ncid_out, dimids_mass(4),&
+           & varid_time_in, varid_time_out, ntimes)
+   endif
 
    ! set up time axis indexing
    itime_beg = time_beg
@@ -542,6 +580,7 @@ program run_pressure_interp
    if (itime_beg <= 0) itime_beg = 1
    if (itime_inc <= 0) itime_inc = 1
 
+#ifdef SKIP
    ! check for time averaging "bounds"
    ! make sure axis is copied (time axis already copied)
    do i=1,NF90_MAX_NAME;  name_time_bounds(i:i) = ' '; enddo
@@ -572,6 +611,9 @@ program run_pressure_interp
    else
       do_time_bounds = .false.
    endif
+#else
+      do_time_bounds = .false.
+#endif
 
    ! define new variables for output file
    !
@@ -658,19 +700,36 @@ program run_pressure_interp
    endif
 
    ! copy variable info for time averaging
-   if (do_time_bounds) then
-      call copy_variable ( ncid_in, trim(name_time_bounds), ncid_out,    &
-         (/dimid_tbnds,dimids_mass(4)/), varid_time_bounds_out )
-   endif
+   if (.not.do_diurn) then
+      if (do_time_bounds) then
+         call copy_variable ( ncid_in, trim(name_time_bounds), ncid_out,    &
+            (/dimid_tbnds,dimids_mass(4)/), varid_time_bounds_out )
+      endif
 
-   if (do_time_average) then
-      do i = 1, 3
-         name_time_average = trim(avg_name)//trim(avg_subnames(i))
-         istat = NF90_INQ_VARID (ncid_in, trim(name_time_average), varid_tavg_in(i))
-         if (istat /= NF90_NOERR) call error_handler (ncode=istat)
-         call copy_variable ( ncid_in, trim(name_time_average), ncid_out,    &
-            (/dimids_mass(4)/), varid_tavg_out(i) )
-      enddo
+      if (do_time_average) then
+         do i = 1, 3
+            name_time_average = trim(avg_name)//trim(avg_subnames(i))
+            istat = NF90_INQ_VARID (ncid_in, trim(name_time_average), varid_tavg_in(i))
+            if (istat /= NF90_NOERR) call error_handler (ncode=istat)
+            call copy_variable ( ncid_in, trim(name_time_average), ncid_out,    &
+               (/dimids_mass(4)/), varid_tavg_out(i) )
+         enddo
+      endif
+   else
+      if (do_time_bounds) then
+         call copy_variable ( ncid_in, trim(name_time_bounds), ncid_out,    &
+         (/dimid_tbnds,dimids_mass(5)/), varid_time_bounds_out )
+      endif
+
+      if (do_time_average) then
+         do i = 1, 3
+            name_time_average = trim(avg_name)//trim(avg_subnames(i))
+            istat = NF90_INQ_VARID (ncid_in, trim(name_time_average), varid_tavg_in(i))
+            if (istat /= NF90_NOERR) call error_handler (ncode=istat)
+            call copy_variable ( ncid_in, trim(name_time_average), ncid_out,    &
+               (/dimids_mass(5)/), varid_tavg_out(i) )
+         enddo
+      endif
    endif
 
    ! define compression parameters
@@ -749,35 +808,76 @@ program run_pressure_interp
 
       ! interpolation setup
 
-      ! surface pressure
-      start(1:3) = (/ is, js, itime /)
-      call read_variable (ncid_in, Fields(index_ps), ps, start=start(1:3))
-      ps(is:ie,js:je,1) = ps(is:ie,js:je,1) * res
+      if (.not.do_diurn) then
+         ! surface pressure
+         start(1:3) = (/ is, js, itime /)
+         call read_variable (ncid_in, Fields(index_ps), ps, start=start(1:3))
+         ps(is:ie,js:je,1) = ps(is:ie,js:je,1) * res
 
 
-      ! mass field interpolation
+         ! mass field interpolation
 
-      ! interpolation initialization for mass fields
-      call interp_field_setup ( ncid_in, ncid_out, bk, pk,&
-           & is, js, itime, otime,&
-           & ps(is:ie,js:je,1),&
-           & zsurf(:,:,1), pout(1:nlev_out) )
+         ! interpolation initialization for mass fields
+         call interp_field_setup ( ncid_in, ncid_out, bk, pk,&
+              & is, js, itime, otime,&
+              & ps(is:ie,js:je,1),&
+              & zsurf(:,:,1), pout(1:nlev_out) )
 
-      ! interpolation of mass fields
-      do i = 1, num_field_names
-         if ( Fields(i)%skip   )    cycle
-         if ( Fields(i)%static )    cycle
-         if ( Fields(i)%y_axis /= nlat ) cycle
+         ! interpolation of mass fields
+         do i = 1, num_field_names
+            if ( Fields(i)%skip   )    cycle
+            if ( Fields(i)%static )    cycle
+            if ( Fields(i)%y_axis /= nlat ) cycle
+
+            kv = max(1,Fields(i)%z_axis)
+            if ( verbose > 1 ) then
+               write (error_unit, *) 'Interp field = ',trim(Fields(i)%name)
+            endif
 
-         kv = max(1,Fields(i)%z_axis)
-         if ( verbose > 1 ) then
-            write (error_unit, *) 'Interp field = ',trim(Fields(i)%name)
-         endif
+            ! interpolation of mass fields
+            call interp_field ( ncid_in, ncid_out, Interp(kv),&
+                 & is, js, itime, otime, Fields(i) )
+         enddo
+      else
+         ! surface pressure
+         do itod = 1,ntod
+            istat =  NF90_GET_VAR (ncid_in, varid_tod_in, tod, start=(/itod/))
+            if (istat /= NF90_NOERR) call error_handler ('getting time coord value', &
+                                                          ncode=istat)
+
+            istat =  NF90_PUT_VAR (ncid_out, varid_tod_out, tod, start=(/itod/))
+            if (istat /= NF90_NOERR) call error_handler ('putting time coord value', &
+                                                          ncode=istat)
+            start(1:4) = (/ is, js, itod, itime /)
+            call read_variable2 (ncid_in, Fields(index_ps), psdiurn, start=start(1:4))
+            psdiurn(is:ie,js:je,1) = psdiurn(is:ie,js:je,1) * res
+
+            ! mass field interpolation
+
+            ! interpolation initialization for mass fields
+            call interp_field_setup2 ( ncid_in, ncid_out, bk, pk,&
+                 & is, js, itime, otime, itod,&
+                 & psdiurn(is:ie,js:je,1),&
+                 & zsurf(:,:,1), pout(1:nlev_out) )
+
+            ! interpolation of mass fields
+            do i = 1, num_field_names
+               if ( Fields(i)%skip   )    cycle
+               if ( Fields(i)%static )    cycle
+               if ( Fields(i)%y_axis /= nlat ) cycle
+
+               kv = max(1,Fields(i)%z_axis)
+               if ( verbose > 1 ) then
+                  write (error_unit, *) 'Interp field = ',trim(Fields(i)%name)
+               endif
+
+               ! interpolation of mass fields
+               call interp_field ( ncid_in, ncid_out, Interp(kv),&
+                    & is, js, itime, otime, Fields(i), itod )
+            enddo
+         enddo
+      endif
 
-         ! interpolation of mass fields
-         call interp_field ( ncid_in, ncid_out, Interp(kv),&
-              & is, js, itime, otime, Fields(i) )
-      enddo
 
       ! free up allocated space
       call pres_interp_free (Interp(1))
@@ -972,43 +1072,231 @@ contains
 
    !#######################################################################
 
-   subroutine interp_field ( ncid_in, ncid_out, Interp, is, js, itime, otime, Field )
-      integer               , intent(in) :: ncid_in, ncid_out
-      type(pres_interp_type), intent(in) :: Interp
-      integer               , intent(in) :: is, js, itime, otime
-      type(field_type)      , intent(in) :: Field
+   subroutine interp_field_setup2 ( ncid_in, ncid_out, bk, pk,&
+        & is, js, itime, otime, itod, ps,&
+        & zsurf, pout )
+
+      integer, intent(in) :: ncid_in, ncid_out
+      real,    intent(in) :: bk(:), pk(:)
+      integer, intent(in) :: is, js, itime, otime, itod
+      real   , intent(in) :: ps(:,:), zsurf(:,:), pout(:)
+
+      real, dimension(size(ps,1),size(ps,2),size(bk(:))-1) ::&
+           & temp_in, sphum_in, mask_in, pfull, hght_in
+      real, dimension(size(ps,1),size(ps,2),size(bk(:))) :: phalf
+      real, dimension(size(ps,1),size(ps,2),size(pout)) ::&
+           & temp_out, hght_out, sphum_out
+      real   , dimension(size(ps,1),size(ps,2)) :: zbot, tbot, pbot, slp
+      integer, dimension(size(ps,1),size(ps,2)) :: kbot
+      character(len=NF90_MAX_NAME) :: name
+      integer :: start(5), kount(5), ndim, ivar, nlev, kb, kr, i, j, k
+      real    :: sig
+
+      do k = 1, size(bk(:))
+         phalf(:,:,k) = pk(k) + bk(k)*ps(:,:)
+      enddo
+      call compute_pres_full (pk, bk, phalf, pfull)
+
+      ! read in temperature data
+      if (do_temp .or. need_temp) then
+         start = (/ is, js, 1, itod, itime /)
+         call read_variable2 (ncid_in, Fields(index_temp), temp_in, start)
+         ! mask
+         call set_up_input_mask (ncid_in, Fields(index_temp), temp_in, mask_in, kbot)
+      endif
+
+      ! read in specific humidity data
+      if (do_sphum .or. need_sphum) then
+         if (sphum_present) then
+            start = (/ is, js, 1, itod, itime /)
+            call read_variable2 (ncid_in, Fields(index_sphum), sphum_in, start)
+            ! mask
+            if (.not.do_temp) &
+               call set_up_input_mask (ncid_in, Fields(index_sphum), sphum_in, mask_in, kbot)
+         else
+            sphum_in = 0.0
+         endif
+      endif
+
+      ! interpolation setup for standard full pressure levels
+      if (do_hght) then
+         if ( verbose > 1 ) write (error_unit, *) 'Computing field = hght'
+         ! geopotential height (in meters)
+         call compute_height ( zsurf, temp_in, sphum_in, pfull, phalf, hght_in, mask_in)
+         ! interpolate temp, sphum, hght
+         if ( verbose > 1 ) then
+            write (error_unit, *) 'Interp field = ',trim(Fields(index_temp)%name)
+            write (error_unit, *) 'Interp field = ',trim(Fields(index_sphum)%name)
+         endif
+         Interp(1) = pres_interp_init ( pfull, pout, kbot,   &
+            tin=temp_in,  tout=temp_out,  &
+            zin=hght_in,  zout=hght_out,  &
+            qin=sphum_in, qout=sphum_out, &
+            use_extrap=.not.mask_extrap   )
+      else if ((do_temp.and.do_sphum).and.(.not.do_hght)) then
+         if ( verbose > 1 ) then
+            write (error_unit, *) 'Interp field = ',trim(Fields(index_temp)%name)
+            write (error_unit, *) 'Interp field = ',trim(Fields(index_sphum)%name)
+         endif
+         ! interpolate temp, sphum
+         Interp(1) = pres_interp_init ( pfull, pout, kbot,   &
+            tin=temp_in,  tout=temp_out,  &
+            qin=sphum_in, qout=sphum_out, &
+            use_extrap=.not.mask_extrap   )
+      else if ((do_temp).and.(.not.do_hght)) then
+         if ( verbose > 1 ) write (error_unit, *) 'Interp field = ',trim(Fields(index_temp)%name)
+         ! interpolate temp
+         Interp(1) = pres_interp_init ( pfull, pout, kbot,   &
+            tin=temp_in,  tout=temp_out,  &
+            use_extrap=.not.mask_extrap   )
+      else if ((do_sphum).and.(.not.do_hght)) then
+         if ( verbose > 1 ) write (error_unit, *) 'Interp field = ',trim(Fields(index_sphum)%name)
+         ! interpolate sphum
+         Interp(1) = pres_interp_init ( pfull, pout, kbot,   &
+            qin=sphum_in, qout=sphum_out, &
+            use_extrap=.not.mask_extrap   )
+      else
+         ! initialize only (need kbot)
+         start = (/ is, js, 1, itod, itime /)
+         call read_and_set_up_mask2 (ncid_in, start, mask_in, kbot)
+         Interp(1) = pres_interp_init ( pfull, pout, kbot,   &
+            use_extrap=.not.mask_extrap   )
+      endif
+
+      ! interpolation setup for standard full pressure levels
+      if (do_phalf) then
+         Interp(2) = pres_interp_init ( phalf, pout, kbot )
+      endif
+
+      ! write out data
+      start = (/  1,  1,  1, itod, otime /)
+
+      ! temperature
+      if (do_temp) then
+         if ( verbose > 1 ) write (error_unit, *) 'Writing field = ',trim(Fields(index_temp)%name)
+         if (Fields(index_temp)%do_miss_out) then
+            call mask_data (Fields(index_temp)%unpkd_missval_out, Interp(1)%mask, temp_out)
+         endif
+         call write_variable2 (ncid_out, Fields(index_temp), temp_out, start)
+      endif
+
+      ! specific humidity
+      if (do_sphum) then
+         if ( verbose > 1 ) write (error_unit, *) 'Writing field = ',trim(Fields(index_sphum)%name)
+         if (Fields(index_sphum)%do_miss_out) then
+            call mask_data (Fields(index_sphum)%unpkd_missval_out, Interp(1)%mask, sphum_out)
+         endif
+         call write_variable2 (ncid_out, Fields(index_sphum), sphum_out, start)
+      endif
+
+      ! height
+      if (do_hght) then
+         if ( verbose > 1 ) write (error_unit, *) 'Writing field = hght'
+         call mask_data (mval_hght, Interp(1)%mask, hght_out)
+         istat = NF90_PUT_VAR (ncid_out, varid_hght_out, hght_out, start)
+         if (istat /= NF90_NOERR) call error_handler ('putting hght data', ncode=istat)
+      endif
+
+      ! sea level pressure
+      if (do_slp) then
+         if ( verbose > 1 ) write (error_unit, *) 'Computing field = slp'
+         do j = 1, size(ps,2)
+            do i = 1, size(ps,1)
+               kb = kbot(i,j)
+               pbot(i,j) = phalf(i,j,kb+1)
+
+               if ( abs(zsurf(i,j)) > 0.0001 ) then
+                  ! get ref level for temp (as in spectral model)
+                  do k = 1, kb
+                     sig = pfull(i,j,k)/pbot(i,j)
+                     if ( sig > 0.8 ) then
+                        kr = k
+                        exit
+                     endif
+                  enddo
+
+                  tbot(i,j) = temp_in(i,j,kr) * sig ** mrgog    ! spectral
+                  slp(i,j) = 0.01 * pbot(i,j) *&
+                       & ( 1.0 + tlapse * zsurf(i,j) / tbot(i,j) ) ** gorg
+                       ! & ( 1.0 + tlapse * zbot(i,j) / tbot(i,j) ) ** gorg
+               else
+                  slp(i,j) = pbot(i,j) * 0.01
+               endif
+            enddo
+         enddo
+
+         if ( verbose > 1 ) write (error_unit, *) 'Writing field = slp'
+         start(1:4) = (/ is, js, itod, otime /)
+         istat = NF90_PUT_VAR (ncid_out, varid_slp_out, slp, start(1:4))
+         if (istat /= NF90_NOERR) call error_handler ('putting slp data', ncode=istat)
+      endif
+   end subroutine interp_field_setup2
+
+!#######################################################################
 
-      real, dimension(Interp%nx,Interp%ny,Interp%kx) :: data_in
-      real, dimension(Interp%nx,Interp%ny,Interp%nz) :: data_out
+   subroutine interp_field ( ncid_in, ncid_out, Interp, is, js, itime, otime, Field, itodarg )
 
-      integer :: start_in(4), start_out(4), istat
+   integer               , intent(in) :: ncid_in, ncid_out
+   type(pres_interp_type), intent(in) :: Interp
+   integer               , intent(in) :: is, js, itime, otime
+   type(field_type)      , intent(in) :: Field
+   integer, intent(in), optional :: itodarg
 
-      ! skip fields have been processed in the initialization routine
-      if (trim(Field%name) == trim(temp_name))  return
-      if (trim(Field%name) == trim(sphum_name)) return
+   real, dimension(Interp%nx,Interp%ny,Interp%kx) :: data_in
+   real, dimension(Interp%nx,Interp%ny,Interp%nz) :: data_out
 
-      ! read in data
+   integer :: start_in(5), start_out(5), istat, itod
 
-      start_in (1:3) = (/ is, js, 1 /)
+!     ---- skip fields have been processed in the initialization routine
+       if (trim(Field%name) == trim(temp_name))  return
+       if (trim(Field%name) == trim(sphum_name)) return
+
+!     ---- read in data ----
+
+       start_in (1:3) = (/ is, js, 1 /)
       !start_out(1:3) = (/ is, js, 1 /)
       start_out(1:3) = (/  1,  1, 1 /)
       start_in (Field%ndim) = itime
       start_out(Field%ndim) = otime
 
       ! 2-d or 3-d data allowed
+      if (present(itodarg)) then
+         itod=itodarg
+      else
+         itod=0
+      endif
       select case ( Field%ndim )
-       case (4)
+       case (5)
          ! check vert axis length
          if ( Field%dimlen(3) /= Interp%kx ) print *, trim(Field%name)
          if ( Field%dimlen(3) /= Interp%kx ) call error_handler (&
               & 'vertical coordinate resolution inconsistent')
-         call read_variable (ncid_in, Field, data_in, start_in(1:4))
+         call read_variable2 (ncid_in, Field, data_in, start_in(1:5))
          ! interpolation
          call pres_interp (Interp, data_in, data_out)
          if (Field%do_miss_out)&
               & call mask_data (Field%unpkd_missval_out, Interp%mask, data_out)
          ! write out data
-         call write_variable (ncid_out, Field, data_out, start_out(1:4))
+         call write_variable2 (ncid_out, Field, data_out, start_out(1:5))
+       case (4)
+         if (itod .gt. 0) then
+           start_in(3)=itod
+           start_out(3)=itod
+           call read_variable2 (ncid_in, Field, data_in(:,:,itod:1), start_in(1:4))
+           call write_variable2 (ncid_out, Field, data_in(:,:,itod:1), start_out(1:4))
+         else
+            ! check vert axis length
+            if ( Field%dimlen(3) /= Interp%kx ) print *, trim(Field%name)
+            if ( Field%dimlen(3) /= Interp%kx ) call error_handler (&
+                 & 'vertical coordinate resolution inconsistent')
+            call read_variable (ncid_in, Field, data_in, start_in(1:4))
+            ! interpolation
+            call pres_interp (Interp, data_in, data_out)
+            if (Field%do_miss_out)&
+                 & call mask_data (Field%unpkd_missval_out, Interp%mask, data_out)
+            ! write out data
+            call write_variable (ncid_out, Field, data_out, start_out(1:4))
+         endif
        case (3)
          call read_variable (ncid_in, Field, data_in(:,:,1:1), start_in(1:3))
          call write_variable (ncid_out, Field, data_in(:,:,1:1), start_out(1:3))
@@ -1147,8 +1435,9 @@ contains
 
    !#######################################################################
 
-   subroutine set_all_field_names ( ncid, field_names, num_field_names )
+   subroutine set_all_field_names ( ncid, do_diurn, field_names, num_field_names )
       integer,          intent(in)    :: ncid
+      logical,          intent(in)    :: do_diurn
       character(len=*), intent(inout) :: field_names(:)
       integer,          intent(inout) :: num_field_names
 
@@ -1179,14 +1468,27 @@ contains
          if ( trim(name) == 'res') cycle    ! for eta coordinate
 
          ! must be 2-d, 3-d, or 4-d field
-         static = dimids(ndim) .ne. rdim
-         threed = .false.
+         if (.not. do_diurn) then
+            static = dimids(ndim) .ne. rdim
+            threed = .false.
+         else
+            static = dimids(ndim) .ne. rdim-2
+            threed = .false.
+         endif
 
          ! skip 3-d fields if flag set
          select case (ndim)
-          case (4)
-            if (static) cycle   ! not possible
+          case (5)
+            if (static) cycle
             threed = .true.
+          case (4)
+            if (.not.do_diurn) then
+               if (static) cycle   ! not possible
+               threed = .true.
+            else
+               if (static) threed =.true.
+               if (.not.static .and. .not.do_all_fields) cycle
+            endif
           case (3)
             !if (.not.static .and. do_all_3d_fields) cycle
             if (static) threed = .true.
@@ -1313,8 +1615,12 @@ contains
          Field%do_miss_in = .true.
          Field%missval_in = fval
       else
-         Field%do_miss_in = .false.
-         Field%missval_in = 0.
+!rjw          Field%do_miss_in = .false.
+!rjw          Field%missval_in = 0.
+         mval= -1.e10
+         fval= -1.e10
+         Field%do_miss_in = .true.
+         Field%missval_in = mval
       endif
 
       ! unpack input missing value (for comparison with unpacked data)
@@ -1333,10 +1639,10 @@ contains
       character(len=NF90_MAX_NAME), intent(in)    :: name
       integer          ,       intent(in)    :: ncid_in, ncid_out
       type(field_type) ,       intent(inout) :: Field
-      logical :: answer
+      logical :: answer,diurn
       integer :: rdim, zcoord, idim
       integer :: i, istat, ipack, ioffs, imiss, ifill
-      integer :: dimids(4)
+      integer :: dimids(5)
       real    :: mval, fval
 
       call get_input_variable_info ( name, ncid_in, Field )
@@ -1375,10 +1681,11 @@ contains
 
       ! set dimension identifiers for output file
       if (Field%dimlen(2) == nlat)   dimids = dimids_mass
+      diurn = dimids(5).gt.0
 
       ! copy input variable metadata to output file
       select case (Field%ndim)
-       case (4)
+       case (5)
          if (Field%dimlen(3) == nlev)   zcoord = 1
          if (Field%dimlen(3) == nlev+1) zcoord = 2
          if (zcoord == 0) return
@@ -1386,6 +1693,37 @@ contains
          call copy_variable ( ncid_in, trim(name), ncid_out,&
               & dimids, Field%varid_out,&
               & missvalue=Field%missval_out  )
+       case (4)
+         if (diurn) then
+            if (Field%static) then
+               if (Field%dimlen(3) == nlev)   zcoord = 1
+               if (Field%dimlen(3) == nlev+1) zcoord = 2
+               if (zcoord == 0) return
+               return               ! no 4d static fields allowed
+               call copy_variable ( ncid_in, trim(name), ncid_out,&
+                    & dimids(1:4), Field%varid_out,&
+                    & missvalue=Field%missval_out )
+            else
+               ! 2d non-static field
+               if (Field%do_miss_in) then
+                  call copy_variable ( ncid_in, trim(name), ncid_out,&
+                       & (/dimids(1),dimids(2),dimids(4),dimids(5)/), Field%varid_out,&
+                       & missvalue=Field%missval_out )
+               else
+                  call copy_variable ( ncid_in, trim(name), ncid_out,&
+                       & (/dimids(1),dimids(2),dimids(4),dimids(5)/), Field%varid_out )
+               endif
+            endif
+
+         else
+            if (Field%dimlen(3) == nlev)   zcoord = 1
+            if (Field%dimlen(3) == nlev+1) zcoord = 2
+            if (zcoord == 0) return
+            if (Field%static) return   ! 4d static fields not allowed
+            call copy_variable ( ncid_in, trim(name), ncid_out,&
+                 & dimids(1:4), Field%varid_out,&
+                 & missvalue=Field%missval_out  )
+         endif
        case (3)
          if (Field%static) then
             if (Field%dimlen(3) == nlev)   zcoord = 1
@@ -1542,6 +1880,53 @@ contains
 
    !#######################################################################
 
+   subroutine read_and_set_up_mask2 (ncid, start, mask, kbot)
+      integer,           intent(in)  :: ncid, start(5)
+      real,              intent(out) :: mask(:,:,:)
+      integer,           intent(out) :: kbot(:,:)
+
+      character(len=NF90_MAX_NAME) :: name
+      real,  dimension(size(mask,1),size(mask,2),size(mask,3)) :: values
+      integer :: ivar, nvar, istat
+      logical :: done
+      type(field_type) :: Field
+
+      done = .false.
+
+      ! get number of variables and time axis id
+      istat = NF90_INQUIRE (ncid, nvariables=nvar)
+      if (istat /= NF90_NOERR) call error_handler (ncode=istat)
+
+      do ivar = 1, nvar
+         istat = NF90_INQUIRE_VARIABLE ( ncid, ivar, name=name )
+         if (istat /= NF90_NOERR) call error_handler (ncode=istat)
+         call get_input_variable_info ( name, ncid, Field )
+
+         ! find first 3d mass field
+         if (Field%ndim <= 2) cycle
+         if (Field%ndim <= 3 .and. .not.Field%static) cycle
+         if (Field%ndim == 4 .and.      Field%static) cycle
+
+         ! must have mass dimension lengths and full vertical levels
+         if (Field%dimlen(1) /= nlon .or. Field%dimlen(2) /= nlat .or. Field%dimlen(3) /= nlev) cycle
+
+         ! must have input missing value
+         if (.not.Field%do_miss_in) cycle
+
+         call read_variable2     (ncid, Field, values, start )
+         call set_up_input_mask (ncid, Field, values, mask, kbot)
+         done = .true.
+         exit
+      enddo
+
+      if (.not.done) then
+         mask = 1.0
+         kbot = nlev
+      endif
+   end subroutine read_and_set_up_mask2
+
+   !#######################################################################
+
    subroutine read_write_static (ncid_in, ncid_out, is, ie, js, je, Field)
       integer         , intent(in) :: ncid_in, ncid_out, is, ie, js, je
       type(field_type), intent(in) :: Field
@@ -1609,6 +1994,56 @@ contains
 
    !#######################################################################
 
+   subroutine read_variable2 ( ncid, Field, values, start )
+      integer,           intent(in)  :: ncid
+      type(field_type),  intent(in)  :: Field
+      real,              intent(out) :: values(:,:,:)
+      integer, optional, intent(in)  :: start(:)
+
+      integer :: ndim, nd, begin(5)
+
+      ! number of dimensions (less the time dimension)
+      ndim = Field%ndim
+      if (.not.Field%static) ndim = ndim-2
+
+      ! start indices
+      nd    = Field%ndim
+      begin = 1
+      if (present(start)) then
+         nd = size(start)
+         begin(1:nd) = start
+      endif
+
+      select case (ndim)
+       case(1)
+         istat = NF90_GET_VAR (ncid, Field%varid_in, values(:,1,1), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('getting 1d var='//trim(Field%name), ncode=istat)
+       case(2)
+         istat = NF90_GET_VAR (ncid, Field%varid_in, values(:,:,1), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('getting 2d var='//trim(Field%name), ncode=istat)
+       case(3)
+         istat = NF90_GET_VAR (ncid, Field%varid_in, values(:,:,:), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('in rv2: getting 3d var='//trim(Field%name), ncode=istat)
+       case default
+         call error_handler ('invalid number of dimensions in read_variable, var='//trim(Field%name))
+      end select
+
+      ! unpack values
+      if (Field%do_pack) then
+         if (Field%do_miss_in) then
+            where (values .ne. Field%missval_in)
+               values = values*Field%pack(1) + Field%pack(2)
+            elsewhere
+               values = Field%unpkd_missval_in
+            endwhere
+         else
+            values = values*Field%pack(1) + Field%pack(2)
+         endif
+      endif
+   end subroutine read_variable2
+
+   !#######################################################################
+
    subroutine write_variable ( ncid, Field, values, start )
       integer,           intent(in)    :: ncid
       type(field_type),  intent(in)    :: Field
@@ -1659,6 +2094,55 @@ contains
 
    !#######################################################################
 
+   subroutine write_variable2 ( ncid, Field, values, start )
+      integer,           intent(in)    :: ncid
+      type(field_type),  intent(in)    :: Field
+      real,              intent(inout) :: values(:,:,:)
+      integer, optional, intent(in)    :: start(:)
+
+      integer :: ndim, nd, begin(5)
+
+      ! pack values (overwrite input data)
+      if (Field%do_pack) then
+         if (Field%do_miss_out) then
+            where (values .ne. Field%unpkd_missval_out)
+               values = nint((values-Field%pack(2))/Field%pack(1))
+            elsewhere
+               values = Field%missval_out
+            endwhere
+         else
+            values = nint((values-Field%pack(2))/Field%pack(1))
+         endif
+      endif
+
+      ! number of dimensions (less the time dimension)
+      ndim = Field%ndim
+      if (.not.Field%static) ndim = ndim-2
+
+      ! start indices
+      nd    = Field%ndim
+      begin = 1
+      if (present(start)) then
+         nd = size(start)
+         begin(1:nd) = start
+      endif
+
+      select case (ndim)
+       case(1)
+         istat = NF90_PUT_VAR (ncid, Field%varid_out, values(:,1,1), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('putting 1d var='//trim(Field%name), ncode=istat)
+       case(2)
+         istat = NF90_PUT_VAR (ncid, Field%varid_out, values(:,:,1), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('putting 2d var='//trim(Field%name), ncode=istat)
+       case(3)
+         istat = NF90_PUT_VAR (ncid, Field%varid_out, values(:,:,:), start=begin(1:nd))
+         if (istat /= NF90_NOERR) call error_handler ('putting 3d var='//trim(Field%name), ncode=istat)
+       case default
+         call error_handler ('invalid number of dimensions in write_variable2, var='//trim(Field%name))
+      end select
+   end subroutine write_variable2
+   !#######################################################################
+
    subroutine compute_pres_full (pk, bk, ph, pf)
       real, intent(in)  :: pk(:), bk(:), ph(:,:,:)
       real, intent(out) :: pf(:,:,:)
